### PR TITLE
Stable Diffusion using aot.export and external parameters

### PR DESCRIPTION
**Pull request by aviator19941**
- Saves weights to .safetensors file
- Load weights at runtime with a "stripped" .mlir

### CODE REVIEW

File: python/turbine_models/custom_models/sd_inference/clip_test.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    "--external_weights",
+    type=str,
+    default=None,
+    help="saves ir/vmfb without global weights for size and readability, options [safetensors]",
+)
+
+prompt = ["a photograph of an astronaut riding a horse"]
+
+
+def save_external_weights(
```
Comment (aviator19941) [inline python/turbine_models/custom_models/sd_inference/clip_test.py:47]: move to utils file

File: python/turbine_models/custom_models/sd_inference/clip_test.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+            target_backends=["llvm-cpu"],
+            extra_args=flags,
+        )
+        with open(f"{safe_name}.vmfb", "wb+") as f:
+            f.write(flatbuffer_blob)
+        print("Saved to", safe_name + ".vmfb")
+        exit()
+
+
+def largest_error(array1, array2):
```
Comment (aviator19941) [inline python/turbine_models/custom_models/sd_inference/clip_test.py:134]: move to utils file

**Comment by aviator19941** [comment]
https://github.com/nod-ai/SHARK-Turbine/issues/152

File: python/turbine_models/custom_models/sd_inference/unet_test.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                    unet_model, external=True, external_scope="", name_mapper=mapper.get
+                )
+            else:
+                params = export_parameters(unet_model)
+
+            def main(
+                self,
+                sample=AbstractTensor(1, 4, 64, 64, dtype=torch.float32),
+                timestep=AbstractTensor(1, dtype=torch.float32),
+                encoder_hidden_states=AbstractTensor(2, 77, 768, dtype=torch.float32),
```
Comment (aviator19941) [inline python/turbine_models/custom_models/sd_inference/unet_test.py:103]: need to find better way to change AbstractTensor size

**Comment by dan-garvey** [comment]
Is this ready to review?

**Comment by aviator19941** [comment]
> Is this ready to review?

@dan-garvey Yes, please review, thanks!

File: python/shark_turbine/importers/fx_importer.py [hunk 4]
```diff
@@ -628,6 +628,12 @@ def _import_torch_op_overload(
         elif target == torch.ops.aten.lift_fresh_copy.out:
             node.target = target = torch.ops.aten.clone.out
             node.args = (node.args[0], None, node.args[1])
+        # TODO: generalize empty memory_format in the future
```
Comment (dan-garvey) [inline python/shark_turbine/importers/fx_importer.py:631]: can you add else case with an explanation?

File: python/turbine_models/custom_models/sd_inference/clip_test.py (hunk 1 above)
Comment (dan-garvey) [reply python/turbine_models/custom_models/sd_inference/clip_test.py]: yeah add as a TODO: I think xida is looking at making this core

File: python/turbine_models/custom_models/sd_inference/unet_test.py (hunk 3 above)
Comment (dan-garvey) [reply python/turbine_models/custom_models/sd_inference/unet_test.py:88]: What do you mean? You can use None for dynamic if needed

File: python/turbine_models/custom_models/sd_inference/unet_test.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            samples, timestep, encoder_hidden_states, return_dict=False
+        )[0]
+        noise_pred_uncond, noise_pred_text = unet_out.chunk(2)
+        noise_pred = noise_pred_uncond + self.guidance_scale * (
+            noise_pred_text - noise_pred_uncond
+        )
+        return noise_pred
+
+
+def save_external_weights(
```
Comment (dan-garvey) [inline python/turbine_models/custom_models/sd_inference/unet_test.py]: probably import this from some shared place

**Review by dan-garvey — CHANGES_REQUESTED** [review]
The test files all do the right thing, but I think warrant some refactoring (we are doing the same with stateless llama I think). Try and deduplicate the code and also add tests to the turbine_models ci

File: python/turbine_models/custom_models/sd_inference/unet_test.py (hunk 3 above)
Comment (aviator19941) [reply python/turbine_models/custom_models/sd_inference/unet_test.py:88]: Hey @dan-garvey, the issue is that for SD 1.4/2.1 the last dimension of encoder_hidden_states is 768/1024, respectively. When I try to use None in the AbstractTensor with the constraint set to `encoder_hidden_states.dynamic_dim(2)`, I get this error:

RuntimeError: a and b must have same reduction dim, but got [s2*s3, s2] X [768, 320].

Comment (dan-garvey) [reply python/turbine_models/custom_models/sd_inference/unet_test.py:88]: I see, yeah unless the encoder hidden state is also dynamic it won't work. In this case I'd just parameterize the last dim based on which model you're instantiating

File: python/shark_turbine/importers/fx_importer.py (hunk 4 above)
Comment (aviator19941) [reply python/shark_turbine/importers/fx_importer.py]: Not sure what the else case would be, but I added an explanation for the case.

**Review by dan-garvey — APPROVED** [review]
Looks good to me, can you add a full e2e test (vmfb and compore to torch result) as a follow up?

File: python/turbine_models/custom_models/sd_inference/utils.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                print("Saved params to", external_weight_file)
+
+
+def largest_error(array1, array2):
+    absolute_diff = np.abs(array1 - array2)
+    max_error = np.max(absolute_diff)
+    return max_error
+
+
+def compile_to_vmfb(module_str, target_backends, safe_name):
```
Comment (IanNod) [inline python/turbine_models/custom_models/sd_inference/utils.py:28]: missing the different backend support that was recently added to stateless_llama

File: python/turbine_models/custom_models/sd_inference/clip.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                name_mapper=mapper.get,
+            )
+        else:
+            params = export_parameters(text_encoder_model)
+
+        def main(self, inp=AbstractTensor(1, 77, dtype=torch.int64)):
+            return jittable(text_encoder_model.forward)(inp)
+
+    import_to = "INPUT" if compile_to == "linalg" else "IMPORT"
+    inst = CompiledClip(context=Context(), import_to=import_to)
```
Comment (IanNod) [inline python/turbine_models/custom_models/sd_inference/clip.py:85]: Do we want to provide the option to do quantization on the matmuls like we are for llama?

**Comment by aviator19941** [comment]
> Looks like we are compiling and running the 3 models (clip, unet, and vae) independently? This won't be able to do a full e2e inference to generate an image from a prompt will it?

No, this won't generate an image from a prompt yet. The 3 models are mainly for checking that we can compile and run them without issues as well as verify the results are similar to torch's results.

**Comment by IanNod** [comment]
> > Looks like we are compiling and running the 3 models (clip, unet, and vae) independently? This won't be able to do a full e2e inference to generate an image from a prompt will it?
> 
> No, this won't generate an image from a prompt yet. The 3 models are mainly for checking that we can compile and run them without issues as well as verify the results are similar to torch's results.

ok, cool. Is there anything holding us back from generating an image, or are we planning on doing that when integrating with the web ui?

**Comment by aviator19941** [comment]
> ok, cool. Is there anything holding us back from generating an image, or are we planning on doing that when integrating with the web ui?

Yep, I was planning on doing that when integrating with the web ui

File: python/turbine_models/custom_models/sd_inference/clip.py (hunk 7 above)
Comment (aviator19941) [reply python/turbine_models/custom_models/sd_inference/clip.py:85]: Not sure if we want to provide that option. I can add it later if needed.

Comment (monorimet) [reply python/turbine_models/custom_models/sd_inference/clip.py:98]: Quantized (int8) SD is a popular request but we don't have a proof-of-concept yet. Can be follow-up.

File: python/turbine_models/custom_models/sd_inference/utils.py (hunk 6 above)
Comment (aviator19941) [reply python/turbine_models/custom_models/sd_inference/utils.py]: added the new backend support

**Review by IanNod — APPROVED** [review]
Looks good to me

**Pull request merged**